sbi/inference/posteriors/sbi_posterior.py

```diff
@@ -157,10 +161,11 @@ def log_prob(
         except AttributeError:
             raise ValueError(f"{self._alg_family} cannot evaluate probabilities.")
 
-        if self._alg_family == "snpe":
-            return log_prob_fn(theta, x, norm_posterior=norm_posterior_snpe)
-        else:
-            return log_prob_fn(theta, x)
+        with torch.set_grad_enabled(track_gradients):
```

Comment:
Have you read about `set_grad_enabled`? It seems to introduce some subtle differences (e.g. use dataset sz vs minibatch sz).

Reply:
where did you read about this? I can not find anything like this?

Quoting from an answer [here](https://discuss.pytorch.org/t/confused-about-set-grad-enabled/38417):
"The torch.set_grad_enabled line of code makes sure to clear the intermediate values for evaluation, which are needed to backpropagate during training, thus saving memory. It’s comparable to the with torch.no_grad() statement but takes a bool value."

Reply:
https://discuss.pytorch.org/t/confused-about-set-grad-enabled/38417/2
> The running_loss will be “de-averaged” by multiplying it with inputs.size(0). Therefore you should divide by the whole dataset length, not the number of batches.

Reply:
I'm pretty sure this is not specific to `set_grad_enabled()`, but rather to how losses are treated in pytorch in general.